File: cloud_tree/main/tree_utils.py
class PartnershipClass:
    def __init__(self, id, *partners):
        self.partners = []
        self.establish_relationship(partners)
        self.children = []
        self.isPlaceholder = False
        self.id = id
    
    def establish_relationship(self, *partners):
        for partner in partners[0]:
            partner.partnership = self
            self.partners.append(partner)
    
    def add_children(self, *ch):
        for child in ch:
            self.children.append(child.partnership)
            child.parent_relationship = self

    def __getitem__(self, key):
        if isinstance(key, int) or isinstance(key, slice): 
            return self.children[key]
        if isinstance(key, str):
            for child in self.children:
                if child.node == key: return child
    
    def __len__(self):
        return len(self.children)
    
    def __str__(self):
        if len(self.partners) == 2:
            return "Partnership: {} and {}".format(self.partners[0], self.partners[1])
        else:
            return "Partnership: {}".format(self.partners[0])

File: cloud_tree/frontend/src/tree_scripts/tree_utils.js
export class Person{
    constructor(name, id){
        this.name = name;
        this.id = id;
        this.partnership = null;
        this.y = null;
        this.x = null;
        this.parent_relationship = null;
    }
}

export class Partnership{
    constructor(partners){
        this.partners = [];
        this.establish_relationship(partners);
        this.children = [];
        this.x = null;
        this.y = null;
        this.id = null;
        this.show = false;
    }

    establish_relationship(partners){
        for(var i = 0; i < partners.length; i++){
            if(partners[i] != null){
                partners[i].partnership = this;
                this.partners.push(partners[i]);
            }
        }
    }

    add_children(children){
        if(typeof children.length !== 'undefined'){
            for(var i = 0; i < children.length; i++){
                this.children.push(children[i].partnership);
                children[i].parent_relationship = this;
            }
        } else {
            this.children.push(children.partnership);
            children.parent_relationship = this;
        }
    }
}

